compiler/rustc_abi/src/layout.rs

```diff
@@ -935,6 +935,18 @@ pub trait LayoutCalculator {
             align = align.min(AbiAndPrefAlign::new(pack));
         }
 
+        let align_mismatch = match abi {
+            Abi::Scalar(x) => x.align(dl).abi != align.abi,
+            Abi::ScalarPair(x, y) => align.abi < cmp::max(x.align(dl).abi, y.align(dl).abi),
+            Abi::Vector { element, .. } => align.abi < element.align(dl).abi,
+            Abi::Aggregate { .. } => false,
+            Abi::Uninhabited => unreachable!(),
+        };
+        if align_mismatch {
+            // mismatched alignment: reset to Aggregate
+            abi = Abi::Aggregate { sized: true };
+        }
```

Comment:
So this still first computes a bad layout but then later patches that up? That doesn't sound like a great approach to me... but I don't know this code. @oli-obk or @eddyb would know better.

Reply:
Yea it's perhaps not ideal. on first pass I did try to do it iteratively but realized we only do the packed alignment adjustment after the fields loop new anyways.

Reply:
right above we do the `repr.pack` check. You could compare the existing alignment against the proposed one and pick the aggregate Abi if they mismatch in there. That way you don't have to look into the abi again, but can just rely on the computed alignment.

Reply:
Yea, just comparing `pack != align.abi` would work there but that'd be stronger than necessary for `ScalarPair` & `Vector` right (at least going off of what `layout_sanity_check` looks for)

Reply:
`layout_sanity_check` is a crude approximation of what invariants we care about -- I wrote it, and I don't really know what the layout invariants are. I would caution against using it as the basis for such things. It is a baseline, guarding against some known invalid layouts, not a reference to hit as closely as possible.